### ISSUE TITLE
Mapper maps parameter names of inner objects

### ISSUE THREAD

**Issue by TimWolla** [CuyZ/Valinor#240]
Unfortunately I was not able to think of a better title. Please consider the following failing test case:

```php
<?php

declare(strict_types=1);

namespace CuyZ\Valinor\Tests\Integration\Mapping\Object;

use CuyZ\Valinor\Mapper\MappingError;
use CuyZ\Valinor\MapperBuilder;
use CuyZ\Valinor\Tests\Integration\IntegrationTest;

final class ParameterNamesTest extends IntegrationTest
{
    /**
     * @dataProvider timezoneTargets
     */
    public function test_wrong_parameter_names_throws_exception(string $class): void
    {
        $source = [
            'timezone' => 'UTC',
        ];

        $this->expectException(MappingError::class);

        (new MapperBuilder())->mapper()->map($class, $source);
    }

    /**
     * @dataProvider timezoneTargets
     */
    public function test_right_parameter_names_work(string $class): void
    {
        $source = [
            'tz1' => 'UTC',
        ];

        self::assertSame('UTC', (new MapperBuilder())->mapper()->map($class, $source)->tz1->getName());
    }

    public function timezoneTargets()
    {
        return [
            [Target1::class],
            [Target2::class],
        ];
    }
}

final class Target1
{
    public function __construct(
        public readonly \DateTimeZone $tz1,
    ) {
    }
}

final class Target2
{
    public function __construct(
        public readonly \DateTimeZone $tz1,
        public readonly ?\DateTimeZone $tz2 = null,
    ) {
    }
}
```

Now consider the following situation:

1. In the initial version of my application the target class is `Target1`. This class expects a single `tz1` parameter containing a value timezone string.
2. The user erroneously sends `timezone=UTC` (i.e. accidentally uses `timezone` instead of `tz1` for the parameter name).
3. This succeeds, I believe because the first parameter of `DateTimeZone::__construct()` is called `$timezone`.
4. In a later version of my application I add a new optional parameter `$tz2`.
5. Now the client in (2) receives an error. The input was invalid before the new parameter was added, but only the addition of the new optional parameter exposed the error.

**Comment by romm** [comment]
Hi @TimWolla thanks for the report, I think this issue should be fixed with https://github.com/CuyZ/Valinor/commit/72cba320f582c7cda63865880a1cbf7ea292d2b1 which is a part of the [version `1.0.0` release](https://github.com/CuyZ/Valinor/releases/tag/1.0.0).

Feel free to re-open if the issue persists. 😉

**Comment by romm** [comment]
Hi @TimWolla, what a fast post-release feedback. 🙂 

I honestly forgot about this usecase, I'm sorry it troubles your application. I'll have to give it some thoughts, maybe a flag should be added in the mapper builder to switch the feature on/off. If you see any better way, please let me know.

In the meantime, I re-open this issue.

**Comment by TimWolla** [comment]
> If you see any better way, please let me know.

My understanding is that for single-parameter classes the following should work:

- Directly providing the single parameter with the appropriate key for the "outer object": `{ "tz1": "Europe/Berlin" }`
- Alternatively providing the single parameter within a nested object: `{ "tz1": { "timezone": "Europe/Berlin" } }` (this is the reason for the revert).

But what should not work:

- Using the single parameter with the key of the "nested object": `{ "timezone": "Europe/Berlin" }` # There is no "timezone" key in `Target1`.

So basically if my updated test passes, the behavior should be correct, it supports the use case in #269, but correctly rejects garbage.

**Comment by TimWolla** [comment]
What *might* be reasonable as well is:

- Directly providing the single parameter: `"Europe/Berlin"` # both `Target1` and the inner `DateTimeZone` only have a single parameter.
